src/utils/vcs.rs

```diff
@@ -427,8 +434,131 @@ pub fn find_heads(
     Ok(rv)
 }
 
+// Get commits from git history upto previous commit.
+// Returns a tuple of Vec<GitCommits> and the `prev_commit` if it exists in the git tree.
+pub fn get_commits_from_git<'a>(
+    repo: &'a Repository,
+    prev_commit: &str,
+) -> Result<(Vec<Commit<'a>>, Option<Commit<'a>>), Error> {
+    let mut revwalk = repo.revwalk()?;
+    revwalk.push_head()?;
+
+    let default_count = 20;
+    let commit_filter = move |id:Result<git2::Oid, git2::Error>| {
+        let id = id.ok()?;
+
+        let commit = repo.find_commit(id).ok()?;
+
+        Some(commit)
+    };
+
+    match git2::Oid::from_str(prev_commit) {
+        Ok(prev) => {
+            let mut found = false;
+            let mut result: Vec<Commit> = revwalk
+                .take_while(|id| match id {
+                    Ok(id) => {
+                        if found {
+                            return false;
+                        }
+                        if prev == *id {
+                            found = true;
+                        }
+                        true
+                    }
+                    _ => true,
+                })
+                .filter_map(commit_filter)
+                .collect();
+            // If there is a previous commit but cannot find it in git history, throw an error.
+            if !found {
+                return Err(format_err!(
+                    "Couldn’t not find the SHA of the previous release in the git history. Increase your git clone depth.",
+                ));
+            }
+            let prev = result.pop();
+            Ok((result, prev))
+        }
+        Err(_) => {
+            // If there is no previous commit, return the default number of commits
+            let mut result: Vec<Commit> = revwalk
+                .take(default_count + 1)
+                .filter_map(commit_filter)
+                .collect();
+
+            if result.len() == default_count + 1 {
+                let prev = result.pop();
+                Ok((result, prev))
+            } else {
+                Ok((result, None))
+            }
+        }
+    }
+}
+
+pub fn generate_patch_set(
+    repo: &Repository,
+    commits: Vec<Commit>,
+    previous: Option<Commit>,
+    repository: &str,
+) -> Result<Vec<GitCommit>, Error> {
+    let mut result = vec![];
+    for (index, commit) in commits.iter().enumerate() {
+        let mut git_commit = GitCommit {
+            id: commit.id().to_string(),
+            author_name: commit.author().name().map(|s| s.to_owned()),
+            author_email: commit.author().email().map(|s| s.to_owned()),
+            message: commit.message().map(|s| s.to_owned()),
+            repository: repository.to_string(),
+            timestamp: get_commit_time(commit.author().when()),
+            patch_set: vec![],
+        };
+
+        let tree1 = commit.tree()?;
+
+        let tree2 = if index + 1 < commits.len() {
```

Comment:
```suggestion
        let tree2 = if commits.len() > index + 1 {
```

I always prefer `>` instead of `<`, easier to read imo